RecoLocalTracker/SiStripClusterizer/plugins/clusterGPU.cu

```diff
@@ -0,0 +1,556 @@
+    bool *__restrict__ trueCluster = clust_data_d->trueCluster_;
+
+    const int tid = threadIdx.x;
+    const int bid = blockIdx.x;
+    const int nthreads = blockDim.x;
+
+    const int i = nthreads * bid + tid;
+
+    if (i < nSeedStripsNC) {
+      const auto index = seedStripsNCIndex[i];
+      const auto chan = channels[index];
+      const auto fed = chanlocs->fedID(chan);
+      const auto channel = chanlocs->fedCh(chan);
+      const auto det = chanlocs->detID(chan);
+      const auto strip = stripId[index];
+      const auto noise_i = conditions->noise(fed, channel, strip);
+
+      auto noiseSquared_i = noise_i * noise_i;
+      float adcSum_i = static_cast<float>(adc[index]);
+
+      // find left boundary
+      auto indexLeft = index;
+      auto testIndex = index - 1;
+      auto size = 1;
+
+      if (testIndex >= 0 && stripId[testIndex] == stripgpu::invStrip) {
+        testIndex -= 2;
+      }
+
+      if (testIndex >= 0) {
+        auto rangeLeft = stripId[indexLeft] - stripId[testIndex] - 1;
+        auto testchan = channels[testIndex];
+        auto testDet = chanlocs->detID(testchan);
+        auto sameDetLeft = det == testDet;
+
+        while (sameDetLeft && rangeLeft >= 0 && rangeLeft <= MaxSequentialHoles && size < ClusterSizeLimit) {
+          testchan = channels[testIndex];
+          const auto testFed = chanlocs->fedID(testchan);
+          const auto testChannel = chanlocs->fedCh(testchan);
+          const auto testStrip = stripId[testIndex];
+          const auto testNoise = conditions->noise(testFed, testChannel, testStrip);
+          const auto testADC = adc[testIndex];
+
+          if (testADC >= static_cast<uint8_t>(testNoise * ChannelThreshold)) {
+            ++size;
+            indexLeft = testIndex;
+            noiseSquared_i += testNoise * testNoise;
+            adcSum_i += static_cast<float>(testADC);
+          }
+          --testIndex;
+          if (testIndex >= 0 && stripId[testIndex] == stripgpu::invStrip) {
+            testIndex -= 2;
+          }
+          if (testIndex >= 0) {
+            rangeLeft = stripId[indexLeft] - stripId[testIndex] - 1;
+            const auto newchan = channels[testIndex];
+            const auto newdet = chanlocs->detID(newchan);
+            sameDetLeft = det == newdet;
+          } else {
+            sameDetLeft = false;
+          }
+        }
+      }
+
+      // find right boundary
+      auto indexRight = index;
+      testIndex = index + 1;
+
+      if (testIndex < nStrips && stripId[testIndex] == stripgpu::invStrip) {
+        testIndex += 2;
+      }
+
+      if (testIndex < nStrips) {
+        auto rangeRight = stripId[testIndex] - stripId[indexRight] - 1;
+        auto testchan = channels[testIndex];
+        auto testDet = chanlocs->detID(testchan);
+        auto sameDetRight = det == testDet;
+
+        while (sameDetRight && rangeRight >= 0 && rangeRight <= MaxSequentialHoles && size < ClusterSizeLimit) {
```

Comment:
I reduced that duplication somewhat by introducing a lambda for some common code.  In principle I could probably reduce the replication further by introducing a "direction" flag, but it would complicate the code and add some extra branches.